### ISSUE TITLE
Problem connecting OpenQA to Keycloak via OpenID

### ISSUE THREAD

**Issue by AnDrEyKa-111** [os-autoinst/openQA#6065]
I have OpenQA 4.6 and I want to set up OpenID Connect authorization on Keycloak 22.0.3. According to the documentation, only the following specification in /etc/openqa/openqa.ini is needed:

[auth]
method = OpenID
[openid]
provider = https://kc.my.domain/realms/master/.well-known/openid-configuration

For OIDC, also need to specify "Client ID" and "Client secret", but how to do this is not specified in the openQA documentation.

When I try to go to the login page https://openqa.my.domain/login, I always get the message "Forbidden". There is an error in the OpenQA log: "Claiming OpenID identity for URL 'https://openqa.my.domain' failed: no_identity_server: Could not determine ID provider from URL".

You can specify one of the OpenID Connect public providers to check this situation, for example:
provider = https://accounts.google.com/.well-known/openid-configuration
And get the same behavior that I observe myself. 

What am I doing wrong and how do I set up OpenID connect correctly?

**Comment by Martchus** [comment]
We also ran into "Could not determine ID provider from URL" with our OpenID provider in the past and it was an [IPv6 related issue](https://open.qa/docs/#_openid_login_times_out).

Otherwise I'm also not sure as we normally use the default OpenID provider or OAuth2. If your provider also supports OAuth2 you could give that a try instead.

For further debugging, the OpenID-related code is in https://github.com/os-autoinst/openQA/blob/master/lib/OpenQA/WebAPI/Auth/OpenID.pm where you might be able to add additional debug printing.

**Comment by Martchus** [comment]
I say OAuth2 is probably easier to resolve.

The error message sounds indeed like some expected parameters are missing. I think the parameter `username` is missing in your case (considering the debug log message). It is expected because you specified `nickname_from = username`. Maybe it makes sense to configure `nickname_from = preferred_username` in your case. I also think the parameter `id` is missing.

**Comment by Martchus** [comment]
So could you make it work with OAuth2? If there were code changes required, feel free to create a PR.

I now realize that you mean OpenID Connect and not OpenID. I think OpenID Connect is not supported by openQA (which uses https://metacpan.org/pod/Net::OpenID::Consumer which only supports OpenID 1 and 2 judging by its CPAN description).

**Comment by AnDrEyKa-111** [comment]
Yes, OAuth2 is working now! The next step is to figure out how to divide users into administrators, operators and regular users through it - is this possible?

I've just see, there is OpenID, and not OpenID Connect. Sorry, it's different things, Keycloak not working with OpenID. Then we'll just use OAuth2.

**Comment by Martchus** [comment]
> Yes, OAuth2 is working now! The next step is to figure out how to divide users into administrators, operators and regular users through it - is this possible?

Only by manually assigning a role in the users menu after the initial login.
